### ISSUE TITLE
AudioDxe: Windows sound does not work after AudioDxe on some systems

### ISSUE THREAD

**Issue by mikebeaton** [acidanthera/bugtracker#1909]
Reported by @aksm-unmei and @antoniomcr96 in #740

**Comment by mikebeaton** [comment]
I'll post the next test efi's here, to keep it a separate issue.

**Comment by mikebeaton** [comment]
Here's a new set, `AudioDxe[15|20|25|30|35|40|45].efi`, [adtests1.zip](https://github.com/acidanthera/bugtracker/files/7795604/adtests1.zip)

Feel free to do a binary search: start half way, with AudioDxe30.efi, if windows sound works go half way up (i.e. to AudioDxe40.efi), if it does not work, go half way down (i.e. to AudioDxe20.efi), then same again with the new one, go up if Windows sound works, down if it does not - you should need to only test 3 of the 7.

**Comment by aksm-unmei** [comment]
> Here's a new set, `AudioDxe[15|20|25|30|35|40|45].efi`, [adtests1.zip](https://github.com/acidanthera/bugtracker/files/7795604/adtests1.zip)
> 
> Feel free to do a binary search: start half way, with AudioDxe30.efi, if windows sound works go half way up (i.e. to AudioDxe40.efi), if it does not work, go half way down (i.e. to AudioDxe20.efi), then same again with the new one, go up if Windows sound works, down if it does not - you should need to only test 3 of the 7.

I have just tested, seems not a good result:
 * AD30: No audio, chime yes.
 * AD35: No audio, chime yes.
 -> Back to half way down
 * AD15, 20, 25: No audio, chime yes.
All of thems the circle spins 4 times as usual.

**Comment by mikebeaton** [comment]
Dammit. I made it wrong of course. I was concentrating so hard on getting one thing right that I got another thing wrong. Sorry for wasting your time! Let me try again.

**Comment by aksm-unmei** [comment]
> Dammit. I made it wrong of course. I was concentrating so hard on getting one thing right that I got another thing wrong. Sorry for wasting your time! Let me try again.

It's ok! I'm free atm.

**Comment by mikebeaton** [comment]
Okay, dump the previous set and try these please:
[adtests1b.zip](https://github.com/acidanthera/bugtracker/files/7795681/adtests1b.zip)

**Comment by aksm-unmei** [comment]
> Okay, dump the previous set and try these please: [adtests1b.zip](https://github.com/acidanthera/bugtracker/files/7795681/adtests1b.zip)

I have just tested half way down. All of them from 15 to 30, no audio and no chime.

**Comment by mikebeaton** [comment]
None of these tests will play the chime, they are doing incomplete audio setup.

I am surprised it fails so early, but all the same (unless I made another mistake manually preparing these), it follows that this one should work (no chime, but yes Windows sound) - can you test please.
[adtests2.zip](https://github.com/acidanthera/bugtracker/files/7795725/adtests2.zip)

**Comment by aksm-unmei** [comment]
oh yes, this one audio works!

**Comment by mikebeaton** [comment]
Aha! We are getting somewhere. Well it is quite a small place which is breaking audio, then. I need to give you some more test versions, but may not be for a few hours till I can come back to this - sorry. :-/

**Comment by aksm-unmei** [comment]
> Aha! We are getting somewhere. Well it is quite a small place which is breaking audio, then. I need to give you some more test versions, but may not be for a few hours till I can come back to this - sorry. :-/

That's ok, feel free to do your main work. I have to go outside now and will be back home in few hours too 😂.
Thank you for your hard work!

**Comment by antoniomcr96** [comment]
I tested audiodxe12 and audiodxe15. The result is the same as @aksm-unmei, so the problem is probably the same.
Audiodxe12 -> audio works, with faster boot time (two spins of the circle)
Audiodxe15 -> audio doesn't work, and boot time becomes slower (about seven-eight spins)
Thanks for your work

**Comment by mikebeaton** [comment]
Really, boot time is slower just because of the extra code in audiodxe15 - surprising - but useful.

It is something in the PCI init code for the device.

Here are two more to test, please [adtests3.zip](https://github.com/acidanthera/bugtracker/files/7795941/adtests3.zip)

**Comment by antoniomcr96** [comment]
Audio works with both the drivers

**Comment by aksm-unmei** [comment]
@antoniomcr96 can you help me for helping mikebeaton test the binaries? I havent been at home yet so I cant test it now. I think our system are same so the result maybe same

**Comment by mikebeaton** [comment]
[adtests4.zip](https://github.com/acidanthera/bugtracker/files/7795991/adtests4.zip) please

**Comment by antoniomcr96** [comment]
Audio doesn’t work in windows with 14.5% build. Unfortunately I have to go outside now, I will be back soon. Thanks again

**Comment by mikebeaton** [comment]
That's narrowed down the exact code which is breaking it. I'll start to look for fixes.

**Comment by aksm-unmei** [comment]
@mikebeaton sorry for late, I've just back home and tested.
Confirmed that with both AD13 and 14, Windows audio yes. But 14.5 audio broken.

**Comment by mikebeaton** [comment]
Here is a possible fix https://github.com/acidanthera/OpenCorePkg/suites/4801196016/artifacts/134704479 - artefacts from just-updated `audio-test` branch. Don't have too much hope, it's just a first try; so if it does not fix yet - as likely! - please provide debug log.

**Comment by aksm-unmei** [comment]
> Here is a possible fix https://github.com/acidanthera/OpenCorePkg/suites/4801196016/artifacts/134704479 - artefacts from just-updated `audio-test` branch. Don't have too much hope, it's just a first try; so if it does not fix yet - as likely! - please provide debug log.

actually, it's worked! Both chime and Windows audio 😂
Here is debug file:
[opencore-Dell-G7-7588-audio-chime-worked.txt](https://github.com/acidanthera/bugtracker/files/7796710/opencore-Dell-G7-7588-audio-chime-worked.txt)

**Comment by antoniomcr96** [comment]
It works!
[opencore-2021-12-31-170555.txt](https://github.com/acidanthera/bugtracker/files/7796722/opencore-2021-12-31-170555.txt)

**Comment by mikebeaton** [comment]
Not that it was a target of the fix, but what happens with the Windows boot delay?

**Comment by antoniomcr96** [comment]
> Not that it was a target of the fix, but what happens with the Windows boot delay?

The main problem for me is that Lenovo logo (boot logo) stays on screen for a bit longer. Subsequently, with audiodxe, chime plays with a black screen that lasts about 2-3 seconds.
Overall, without audiodxe it takes 10 seconds to reach windows spinning circle.
With audiodxe it takes 15-16 seconds

**Comment by aksm-unmei** [comment]
> Not that it was a target of the fix, but what happens with the Windows boot delay?

mine booting to Windows is normal, fully spins 1.5 rotations then desktop shows up.
But like antoniomcr96 said, there is a small delay time if chime plays successful. After BIOS logo, takes about ~3-5s, then chime plays and OpenCanopy shows up. If there is not chime sound like old test binaries, or AD is disabled, OpenCanopy shows immediately after BIOS logo.

**Comment by mikebeaton** [comment]
Okay, that's all good, as far as it goes - there are known issues of AudioDxe being too slow to start up. Something separate to look at. But the delay once Windows is selected is the same problem as lack of Windows sound, at least - as hoped.

**Comment by antoniomcr96** [comment]
I didn’t understand well what you asked before, sorry. Yes, I confirm that your fix solves also the slower windows boot time AFTER POST 👍👍

**Comment by mikebeaton** [comment]
@tarutaru9250 @antoniomcr96 The fix I created for you breaks sound in macOS on some other systems https://github.com/acidanthera/bugtracker/issues/1945 . The latest commit https://github.com/acidanthera/OpenCorePkg/commit/0455326487726e3ac974f7ea14030accb736ba1d will have broken sound for you in Windows again by default, but please check and confirm that if you add `--restore-nosnoop` as a string the the AudioDxe driver arguments then your fix (now opt-in) still works.

**Comment by tarutaru9250** [comment]
> @tarutaru9250 @antoniomcr96 The fix I created for you breaks sound in macOS on some other systems #1945 . The latest commit [acidanthera/OpenCorePkg@0455326](https://github.com/acidanthera/OpenCorePkg/commit/0455326487726e3ac974f7ea14030accb736ba1d) will have broken sound for you in Windows again by default, but please check and confirm that if you add `--restore-nosnoop` as a string the the AudioDxe driver arguments then your fix (now opt-in) still works.

Thank you very much. I will try it asap.

**Comment by tarutaru9250** [comment]
@mikebeaton confirmed on my Dell, Windows sound is fine with new AudioDxe build and the argument. Thanks for your hard work.
![image](https://user-images.githubusercontent.com/59220499/153353394-fa7d1b6d-db7d-463b-acb2-8607726d0d8a.png)

**Comment by antoniomcr96** [comment]
Everything ok. Thanks
